File: quick_sort_1.java
import java.io.*;
import java.util.*;

public class quick_sort_1 {

    static class Data {
        long key;
        String val;

        Data(long key, String val) {
            this.key = key;
            this.val = val;
        }
    }

    private static Random rand = new Random();

    public static void quickSort(Data[] arr) {
        quickSort(arr, 0, arr.length - 1);
    }

    private static void quickSort(Data[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = randomizedPartition(arr, low, high);
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    private static int randomizedPartition(Data[] arr, int low, int high) {
        int pivotIndex = low + rand.nextInt(high - low + 1);
        swap(arr, pivotIndex, high);
        return partition(arr, low, high);
    }

    private static int partition(Data[] arr, int low, int high) {
        long pivot = arr[high].key;
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j].key <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    private static void swap(Data[] arr, int i, int j) {
        Data temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // Read CSV into Data[]
    public static Data[] readCSV(String filename) throws IOException {
        List<Data> dataList = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                // Assume CSV with 2 columns separated by comma
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    long key = Long.parseLong(parts[0].trim());
                    String val = parts[1].trim();
                    dataList.add(new Data(key, val));
                }
            }
        }
        return dataList.toArray(new Data[0]);
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java QuickSortCSV <csv-file-path>");
            return;
        }

        try {
            // Read CSV file
            Data[] dataArray = readCSV(args[0]);

            // Sort the array
            quickSort(dataArray);

            // Print first 20 sorted records
            System.out.println("First 20 sorted entries:");
            for (int i = 0; i < Math.min(20, dataArray.length); i++) {
                System.out.println(dataArray[i].key + ", " + dataArray[i].val);
            }

        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        } catch (NumberFormatException e) {
            System.err.println("Error parsing key as long integer: " + e.getMessage());
        }
    }
}

File: quick_sort_1.py
import random
import time

class Data:
    def __init__(self, key, val):
        self.key = key
        self.val = val

def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

def partition(arr, low, high):
    pivot = arr[high].key
    i = low - 1
    for j in range(low, high):
        if arr[j].key <= pivot:
            i += 1
            swap(arr, i, j)
    swap(arr, i + 1, high)
    return i + 1

def randomized_partition(arr, low, high):
    pivot_index = random.randint(low, high)
    swap(arr, pivot_index, high)
    return partition(arr, low, high)

def quick_sort(arr, low, high):
    if low < high:
        pi = randomized_partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def quick_sort_wrapper(arr):
    quick_sort(arr, 0, len(arr) - 1)

def read_csv(filename):
    data_list = []
    with open(filename, 'r') as f:
        for line in f:
            parts = line.strip().split(',', 1)
            if len(parts) == 2:
                try:
                    key = int(parts[0].strip())
                    val = parts[1].strip()
                    data_list.append(Data(key, val))
                except ValueError:
                    pass
    return data_list

def write_csv(filename, data):
    with open(filename, 'w') as f:
        for d in data:
            f.write(f"{d.key},{d.val}\n")

def main():
    input_file = 'dataset_1000000.csv'
    output_file = 'quick_sort_1_1000000.csv'

    data = read_csv(input_file)
    input_size = len(data)
    if input_size == 0:
        print("No valid data to sort.")
        return

    start_time = time.time()
    quick_sort_wrapper(data)
    end_time = time.time()

    time_taken_ms = int((end_time - start_time) * 1000)

    write_csv(output_file, data)

    print("Sorting completed.")
    print(f"Input size: {input_size}")
    print(f"Time taken (ms): {time_taken_ms}")
    print(f"Sorted data written to {output_file}")

if __name__ == "__main__":
    main()
